"""
MainWindowEx
Extended MainWindow v·ªõi logic qu·∫£n l√Ω tabs v√† ph√¢n quy·ªÅn
"""
import sys
from pathlib import Path
from PyQt6.QtWidgets import QMainWindow, QVBoxLayout
from PyQt6.QtCore import pyqtSignal

project_root = Path(__file__).resolve().parent.parent
sys.path.insert(0, str(project_root))

from ui.MainWindow import Ui_MainWindow
from ui.PredictionTabWidget import PredictionTabWidget
from ui.DashboardTabWidget import DashboardTabWidget
from models.user import User
from database.connector import DatabaseConnector
from config.database_config import DatabaseConfig
from services.query_service import QueryService


class MainWindowEx(QMainWindow):
    """
    Extended MainWindow v·ªõi logic ch√≠nh
    - Qu·∫£n l√Ω tabs
    - Ph√¢n quy·ªÅn theo role
    - X·ª≠ l√Ω logout
    """
    
    # Signal emit khi logout
    logout_signal = pyqtSignal()
    
    def __init__(self, user: User):
        """
        Kh·ªüi t·∫°o MainWindow
        
        Args:
            user: User ƒë√£ ƒëƒÉng nh·∫≠p
        """
        super().__init__()
        self.user = user
        self.ui = Ui_MainWindow()
        self.ui.setupUi(self)
        
        # Init database connection
        self.db_config = DatabaseConfig.default()
        self.db_connector = DatabaseConnector(self.db_config)
        self.db_connector.connect()
        
        # Init services
        self.query_service = QueryService(self.db_connector)
        
        # Setup UI
        self.setup_user_info()
        self.setup_tabs()
        self.setup_role_permissions()
        
        # Connect signals
        self.ui.btnLogout.clicked.connect(self.handle_logout)
    
    def setup_user_info(self):
        """Hi·ªÉn th·ªã th√¥ng tin user"""
        self.ui.lblWelcome.setText(f"Xin ch√†o, {self.user.username} ({self.user.role})")
    
    def setup_tabs(self):
        """Thi·∫øt l·∫≠p n·ªôi dung c√°c tabs"""
        # Clear default tabs
        self.ui.tabWidget.clear()
        
        # Tab 1: D·ª± B√°o R·ªßi Ro
        self.prediction_widget = PredictionTabWidget(self.query_service)
        self.ui.tabWidget.addTab(self.prediction_widget, "üìä D·ª± B√°o R·ªßi Ro")
        
        # Tab 2: Dashboard
        self.dashboard_widget = DashboardTabWidget()
        self.ui.tabWidget.addTab(self.dashboard_widget, "üìà Dashboard")
    
    def setup_role_permissions(self):
        """
        Thi·∫øt l·∫≠p ph√¢n quy·ªÅn theo role
        
        Rules:
        - Admin: Xem t·∫•t c·∫£ tabs
        - Technical: Xem Prediction + Dashboard
        - Secretary: Ch·ªâ xem Prediction
        """
        if self.user.role == 'Admin':
            # Admin c√≥ quy·ªÅn t·∫•t c·∫£
            print(f"‚úì Admin access: All tabs enabled")
        
        elif self.user.role == 'Technical':
            # Technical c√≥ quy·ªÅn Prediction + Dashboard
            print(f"‚úì Technical access: Prediction + Dashboard enabled")
        
        elif self.user.role == 'Secretary':
            # Secretary ch·ªâ c√≥ quy·ªÅn Prediction
            # ·∫®n tab Dashboard
            dashboard_index = None
            for i in range(self.ui.tabWidget.count()):
                if 'Dashboard' in self.ui.tabWidget.tabText(i):
                    dashboard_index = i
                    break
            
            if dashboard_index is not None:
                self.ui.tabWidget.removeTab(dashboard_index)
            
            print(f"‚úì Secretary access: Only Prediction tab enabled")
        
        else:
            # Unknown role - restrict to Prediction only
            print(f"‚ö† Unknown role '{self.user.role}': Default to Prediction only")
    
    def handle_logout(self):
        """X·ª≠ l√Ω s·ª± ki·ªán logout"""
        print(f"ƒêƒÉng xu·∫•t user: {self.user.username}")
        
        # Close database connection
        if self.db_connector:
            self.db_connector.close()
        
        # Hide window
        self.hide()
        
        # Emit signal
        self.logout_signal.emit()
    
    def closeEvent(self, event):
        """Override closeEvent"""
        if self.db_connector:
            self.db_connector.close()
        event.accept()
